### PR TITLE
Improve wildcard handling for `imgmount` file arguments

### PR DESCRIPTION
**Pull request by kcgen**
Previously, alpha-numeric sorting was used when mounting disk images with wildcards.

This works fine when disk image numbers are zero-padded up to the largest, for example in an 11 disk set: `disk01.img`, `disk02.img`, ..., `disk11.img`.

However, sometimes disks aren't zero-padded resulting in unwanted ordering:

This PR adds natural sorting:

And nicer logging of the result.

- Singles: 

- Pairs: 

- Three or more: 

It also adds filename to the console log messages, which can help when reporting issues or when parsing with a script:

```
2023-02-08 | FAT: Mounted disk1.img as FAT12 volume with 2847 clusters
2023-02-08 | FAT: Mounted disk2.img as FAT12 volume with 2847 clusters
2023-02-08 | FAT: Mounted disk3.img as FAT12 volume with 2847 clusters
2023-02-08 | FAT: Mounted disk4.img as FAT12 volume with 2847 clusters
2023-02-08 | FAT: Mounted disk5.img as FAT12 volume with 2847 clusters
2023-02-08 | FAT: Mounted disk6.img as FAT12 volume with 2847 clusters
2023-02-08 | FAT: Mounted disk7.img as FAT12 volume with 2847 clusters
2023-02-08 | FAT: Mounted disk8.img as FAT12 volume with 2847 clusters
2023-02-08 | FAT: Mounted disk9.img as FAT12 volume with 2847 clusters
2023-02-08 | FAT: Mounted disk10.img as FAT12 volume with 2847 clusters
2023-02-08 | FAT: Mounted disk11.img as FAT12 volume with 2847 clusters
```

Thanks to @Eulisker from the eXoDOS project for catching this bug :+1: !

### CODE REVIEW

File: src/dos/program_imgmount.cpp [hunk 1]
```diff
@@ -235,10 +235,15 @@ void IMGMOUNT::Run(void)
 		return;
 	}
 
+	auto used_wildcards = false;
```
Comment (johnnovak) [inline src/dos/program_imgmount.cpp:238]: `use_wildcards`? (present)

File: src/misc/string_utils.cpp [hunk 2]
```diff
@@ -180,12 +180,61 @@ std::vector<std::string> split(const std::string &seq)
 	return words;
 }
 
+std::string join_with_commas(const std::vector<std::string>& items)
+{
+	// skip the empties: we don't want to see them
+	auto is_not_empty = [](const std::string& s) { return !s.empty(); };
+	const auto num_items = std::count_if(items.begin(), items.end(), is_not_empty);
```
Comment (johnnovak) [inline src/misc/string_utils.cpp:187]: Kinda okay in this case, but you can compose functions better if one function does one thing. So stripping could be extracted into a `strip_empty_strings` function or something, then you can compose them. Having said that, if you use this function a lot, that becomes tiresome...

File: src/misc/string_utils.cpp [hunk 3]
```diff
@@ -180,12 +180,61 @@ std::vector<std::string> split(const std::string &seq)
...
-char *strip_word(char *&line)
+bool natural_compare(const std::string& a_str, const std::string& b_str)
+{
+	auto parse_num = [](auto& it, const auto it_end) {
+		int num = 0;
+		while (it != it_end && isdigit(*it)) {
+			num = num * 10 + (*it - '0');
+			++it;
+		}
+		return num;
```
Comment (johnnovak) [inline src/misc/string_utils.cpp:217]: That's clever! 😄

File: include/string_utils.h [hunk 4]
```diff
@@ -244,7 +244,17 @@ constexpr bool iequals(T1&& a, T2&& b)
...
 
-char *strip_word(char *&cmd);
+// Performs a "natural" comparison between A and B, which is case-insensitive
+// and treats number sequenences as whole numbers. Returns true if A < B. This
+// function can be used with higher order sort rountines, like std::sort.
+//
+// Examples:
+// - ("abc_2", "ABC_10") -> true, because abc_ matches and 2 < 10.
+// - ("xyz_2", "ABC_10") -> false, because 'x' > 'a'.
+// - ("abc123", "abc123=") -> true, simply because the first is shorter.
```
Comment (johnnovak) [inline include/string_utils.h:254]: So what about `("Z", "abc123=")`? In this case you'd need to fall back to the default alphanumeric sort, I think.

I think how it should work is as follows:

- If the two strings have the same prefix (without taking casing into account) and they in numbers, then do the natural sort thing
- In any other scenario, default to alphanumeric sort

**Comment by johnnovak** [comment]
So this log message reads fine:

```
FAT: Mounted disk4.img as FAT12 volume with 2847 clusters
```

But these I find weird:



Drive A mounted? Should be the other way around, no?

```
disk1.img, disk2.img, and disk3.img mounted as drive A
```

File: src/dos/program_imgmount.cpp (hunk 1 above)
Comment (kcgen) [reply src/dos/program_imgmount.cpp:238]: How about `has_wildcards`?  (it's past tense; we're checking if the user used wildcards in their CLI statement.)

I think `has_` is a good fit w/ other bools.

File: src/misc/string_utils.cpp (hunk 2 above)
Comment (kcgen) [reply src/misc/string_utils.cpp:187]: Indeed; hmm.. I included this bit only defensively knowing that /maybe/ someone might pass it vectors with empty strings (and they would get output like `", , and ."` 

On second thought, I think it's even cleaner that we expect inbound vectors to have no empties, so I will assert that in the loop. (And the filenames being added by imgmount and certainly non-empty because they've passed various checks before hand).

File: include/string_utils.h (hunk 4 above)
Comment (kcgen) [reply include/string_utils.h:254]: > If the two strings have the same prefix (without taking casing into account) and they in numbers, then do the natural sort thing
> In any other scenario, default to alphanumeric sort

:+1: that's how it works.

Here's a fun (and contrived!) example of a studio in the middle of development. 

They've got different people working on the intro, game data, ending sequence (currently two drafts remaining), and outro.

Each team shares the floppy disk images:



The game testers mount the intro, data, ending, and outro in that order.

Here's baseline behavior:





(yes, some other bug is blocking the second wildcard set from being handled)

And the PR:

Comment (kcgen) [reply include/string_utils.h:254]: > Drive A mounted? Should be the other way around, no?
> 
> ```
> disk1.img, disk2.img, and disk3.img mounted as drive A
> ```

I agree.. maybe they wanted the letter to be shown first (it stands out a bit more).  but I agree, yours is more natural and preferred.

I swapped the drive letter and string:

```c++
MSG_Add("PROGRAM_MOUNT_STATUS_2","Drive %c is mounted as %s\n");

to 

MSG_Add("PROGRAM_MOUNT_STATUS_2","Images %s mounted as drive %c\n");
```

And it segfaults (because now the arguments need to be swapped) in these usage points:

```
src/dos/program_imgmount.cpp
src/dos/program_mount.cpp        
src/dos/program_mount_common.cpp
```

That's not a big deal - but this is worse.. all the foreign strings for `PROGRAM_MOUNT_STATUS_2` need to be reversed:

contrib/resources/translations/utf-8/de.txt::PROGRAM_MOUNT_STATUS_2
contrib/resources/translations/utf-8/en.txt::PROGRAM_MOUNT_STATUS_2
contrib/resources/translations/utf-8/es.txt::PROGRAM_MOUNT_STATUS_2
contrib/resources/translations/utf-8/fr.txt::PROGRAM_MOUNT_STATUS_2
contrib/resources/translations/utf-8/it.txt::PROGRAM_MOUNT_STATUS_2
contrib/resources/translations/utf-8/nl.txt::PROGRAM_MOUNT_STATUS_2
contrib/resources/translations/utf-8/pl.txt::PROGRAM_MOUNT_STATUS_2
contrib/resources/translations/utf-8/ru.txt::PROGRAM_MOUNT_STATUS_2

So I'd either have to hack those strings or ask the translators to fix them (all)? Unfortunately this is feeling too stuck right now :sob:

**Comment by rderooy** [comment]
> The bigger hit is that all the foreign strings (for `PROGRAM_MOUNT_STATUS_2`) would need to be flipped as well (otherwise those users will segfault):
> 
> * contrib/resources/translations/utf-8/de.txt::PROGRAM_MOUNT_STATUS_2
> * contrib/resources/translations/utf-8/en.txt::PROGRAM_MOUNT_STATUS_2
> * contrib/resources/translations/utf-8/es.txt::PROGRAM_MOUNT_STATUS_2
> * contrib/resources/translations/utf-8/fr.txt::PROGRAM_MOUNT_STATUS_2
> * contrib/resources/translations/utf-8/it.txt::PROGRAM_MOUNT_STATUS_2
> * contrib/resources/translations/utf-8/nl.txt::PROGRAM_MOUNT_STATUS_2
> * contrib/resources/translations/utf-8/pl.txt::PROGRAM_MOUNT_STATUS_2
> * contrib/resources/translations/utf-8/ru.txt::PROGRAM_MOUNT_STATUS_2
> 
> I'd either have to hack those strings (and do a bad translation job) or ask the translators to fix them. Unfortunately this is feeling too stuck right now for this PR - might have to forgo this change sob

**Dutch**
Old:
```
:PROGRAM_MOUNT_STATUS_2
Schijf %c is gekoppeld als %s
```
New:
```
:PROGRAM_MOUNT_STATUS_2
Image %s gekoppeld als schijf %c
```

**German**
Old:
```
:PROGRAM_MOUNT_STATUS_2
Laufwerk %c ist eingebunden als %s
```
New:
```
:PROGRAM_MOUNT_STATUS_2
Image %s ist eingebunden als Laufwerk %c
```

**Comment by kcgen** [comment]
@rderooy : one man translation team! Awesome.

Will go ahead with this :-)

**Comment by rderooy** [comment]
> * PROGRAM_MOUNT_STATUS_2

Thanks ;-)

I'm less fluent in French and Italian, but I think this should be correct...

Original French
```
Lecteur %c monté en tant que %s
```
Updated
```
Image %s monté  en tant que lecteur %c
```

Original Italian
```
Unità %c montata come %s
```
```
Immagine %s montata come unità %c
```

**Comment by kcgen** [comment]
@rderooy : I realized we need a list conjunction word for these languages too.

Apple, banana, **_and_** pear.

**Comment by rderooy** [comment]
Spanish, possibly. Close enough to Italian. Polish and Russian, no way ;-)

**Comment by rderooy** [comment]
Conjunction words:

Dutch: en
German: und
French: et
Italian: e
Spanish: y

**Comment by rderooy** [comment]
I think the Spanish translation would be
```
Imagen %s montada como unidad %c
```

**Comment by FeralChild64** [comment]
Conjunction words:

Polish: i

Translation - old:

```
:PROGRAM_MOUNT_STATUS_2
Dysk %c jest zamontowany jako '%s'.
```

Translation - new:

```
:PROGRAM_MOUNT_STATUS_2
Obraz %s zamontowany jako dysk %c.
```

I'll need to review the Polish translation once again, there are probably more cases where the single quotation marks or the verbs should be removed (like in this case).

**Comment by kcgen** [comment]
@GranMinigun

help needed to translate "and" conjunction to Russian:

So far, I've got `и` (versus `а`, which contrasts)?

File: src/dos/program_imgmount.cpp (hunk 1 above)
Comment (johnnovak) [reply src/dos/program_imgmount.cpp]: Yes "has" is the best, agreed.

File: include/string_utils.h (hunk 4 above)
Comment (johnnovak) [reply include/string_utils.h:254]: Looks good, I think I slightly misunderstood it then how it works 😄

**Comment by kcgen** [comment]
@johnnovak , @rderooy , @FeralChild64 , @GranMinigun , @NicknineTheEagle , @Kappa971 , @Burrito78 

Here are the updated mount status logs, do they look OK? 

Shout if you want to apply any tweaks.

**Comment by johnnovak** [comment]
> @johnnovak , @rderooy , @FeralChild64 , @GranMinigun , @NicknineTheEagle , @Kappa971 , @Burrito78
> 
> Here are the updated mount status logs, do they look OK?
> 
> Regarding "_drive C_ versus _C drive_", the latter is more correct so I've used that for English and Dutch, as my searches found those were the most prevalent and acceptable.
> 
> My searches in the other languages show the "drive C" prefix-approach is more common. Please shout if I got this wrong or if there are more tweaks.
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 

Looks good to me 👍🏻 Minor comment: there's a period at the end of the Polish (?) translation only. Probably should be removed for consistency?

**Comment by NicknineTheEagle** [comment]
Russian text should say "Образ" instead of "Привод".

**Comment by kcgen** [comment]
> Looks good to me 👍🏻 Minor comment: there's a period at the end of the Polish (?) translation only. Probably should be removed for consistency?

I think @FeralChild64 wanted it; but will let him decide.

**Comment by kcgen** [comment]
> Russian text should say "Образ FAT" instead of "Привод FAT".



added.

**Comment by GranMinigun** [comment]
Ideally the comma before "и" should be dropped (Oxford comma is not a thing in Russian), but I'm nitpicking. I also see that everything is in singular form... gettext, so close yet so far away.

**Comment by Kappa971** [comment]
> 

Seems ok to me 👍🏻

**Comment by Burrito78** [comment]
> The Oxford comma did not actually originate at Oxford University in England. One can trace its origin in English guides from the early 20th century. The Oxford comma is "correct" in American Standard English but does not exist in other languages, nor is it mandatory in British or International English.
https://libguides.framingham.edu/c.php?g=894986&p=7278665

Also please change "FAT image" -> "FAT-Image" for german.

Thanks, looking good otherwise!

**Comment by johnnovak** [comment]
> > The Oxford comma did not actually originate at Oxford University in England. One can trace its origin in English guides from the early 20th century. The Oxford comma is "correct" in American Standard English but does not exist in other languages, nor is it mandatory in British or International English.
> > https://libguides.framingham.edu/c.php?g=894986&p=7278665
> 
> Also please change "FAT image" -> "FAT-Image" for german.
> 
> Thanks, looking good otherwise!

Can you even call a filesystem FAT these days? Is that politically correct? Shall we rename it to LargeSized? :trollface:

**Comment by FeralChild64** [comment]
@kcgen @johnnovak You can remove period at the end of sequence in Polish translation if the original string does not have it. I’m going to check the remaining strings for consistency later.

**Comment by kcgen** [comment]
@Burrito78 



@FeralChild64 



I think that does it.  Merging. Huge team effort - thanks everyone!

**Comment by johnnovak** [comment]
> > https://libguides.framingham.edu/c.php?g=894986&p=7278665
> 
> Great reference @Burrito78 , especially those examples.
> 
> I took an engineering technical writing course (long ago), but the last comma was mandatory to differentiate the last "item" exactly like those examples show (and the confusion and legal troubles it can cause without it!)
> 
> This style also makes bullets and commas behave consistently (imagine stuffing the last two items on the same bullet!)
> 
>     * A
> 
>     * B
> 
>     * C
> 
>     * D

I'm a big fan of the Oxford comma too, and so are some of my Aussie colleagues. As @kcgen said, it removes ambiguity which is always a good thing. In general, I'm a comma maximalist! I use as much punctuation as I can; I see them as a courtesy to my reader to help them with parsing my sentences more easily.

**Comment by Burrito78** [comment]
https://www.germanwithnicole.com/blog/36237-sorry-there-is-no-oxford-comma-in-german

Nicole said it best. 

> So be sure to review your German texts when you've written a list und entfernen Sie das Oxford-Komma!

**Comment by johnnovak** [comment]
One the importance of commas; this is a real example from Hungarian history which is somewhat hard to translate, but it illustrates the concept.

TL;DR the queen was assassinated in the end, and the writer of the message wanted to discourage certain parties from the murder, but it can be actually interpreted both ways.

The original message went like this (without commas):

"Kill Queen you must not fear will be good if all agree I do not oppose"

And the two possible interpretations (the author of the messages claimed the second interpretation was correct, and was hence spared from execution):

"Kill Queen, you must not fear, will be good if all agree, I do not oppose"

"Kill Queen you must not, fear will be good, if all agree I do not, oppose"
